### PR TITLE
feat(components): create rating web-component

### PR DESCRIPTION
**Pull request by davidritter-dotcom**

### CODE REVIEW

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+  /**
+   * The number of stars in the rating
+   */
+  @Prop() readonly disabled?: boolean = false;
+
+  /**
+   * The number of stars in the rating
+   */
+  @Prop() readonly readonly?: boolean = false;
```
Comment (b1aserlu) [inline packages/components/src/components/post-rating/post-rating.tsx:33]: These props all have the same description.

**Review by b1aserlu — CHANGES_REQUESTED** [review]

**Review by b1aserlu — APPROVED** [review]

**Comment by davidritter-dotcom** [comment]
One e2e test is failing, but locally the test works. So the reviewer could maybe test if it also works locally on his machine.

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        <svg
+          viewBox="0 0 24 24"
+          xmlns="http://www.w3.org/2000/svg"
+          class={`${this.getClasses(i).join(' ')}`}
+          onClick={this.isInteractive() && (() => this.handleClick(i))}
+          onMouseEnter={this.isInteractive() && (() => this.handleHover(i))}
+          onMouseLeave={this.isInteractive() && (() => this.reset())}
+          cursor={this.isInteractive() ? 'pointer' : 'default'}
+        >
+          <path d="M15.2047 8.01289L15.3173 8.25303L15.5793 8.29449L22.981 9.46594L17.8102 14.8955L17.6402 15.0741L17.6783 15.3177L18.8923 23.0722L12.3555 19.4823L12.1149 19.3501L11.8742 19.4823L5.3374 23.0722L6.55141 15.3177L6.59048 15.0681L6.4128 14.8886L1.0518 9.47202L8.30071 8.41511L8.56513 8.37655L8.68001 8.1353L11.9965 1.17035L15.2047 8.01289Z" />
```
[the remark this answers is not in the document]
Comment (davidritter-dotcom) [reply packages/components/src/components/post-rating/post-rating.tsx:163]: But is it possible to make a border or an outline or something like that on the star in the `post-icon`?

[the remark this answers is not in the document]
Comment (alizedebray) [reply packages/components/src/components/post-rating/post-rating.tsx:154]: No it is not possible to set a border or a stroke on the icons, however you could put 2062 on top of 2574 which would then allow half-filled stars: 
image

But let's discuss this topic during the roundtable!

File: packages/components/cypress/e2e/rating.cy.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    });
+
+    // Unfortunately the test for the hovered state does not always work.
+    //
+    // it('should change appearance on hover', () => {
+    //   cy.get('@stars')
+    //     .its('length')
+    //     .then(length => {
+    //       cy.get('@stars').each(($star, i) => {
+    //         cy.wrap($star).trigger('mouseenter');
```
Comment (oliverschuerch) [inline packages/components/cypress/e2e/rating.cy.ts:34]: What if you try the `mouseover` event?
The `mouseover` event triggers when the mouse pointer enters an element, and its child elements. The `mouseenter` event is only triggered when the mouse pointer enters the an element, but not if it enters one of this elements child elements.

File: packages/components/src/components/post-rating/post-rating.scss [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    outline: none;
+  }
+  &[aria-readonly='true'] .star-container {
+    cursor: default;
+  }
+}
+
+.rating:focus-visible {
+  outline-offset: post.$input-focus-outline-thickness;
+  outline: post.$input-focus-outline-thickness solid post.$input-focus-color;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.scss:23]:
```suggestion
  outline: post.$input-focus-outline-thickness solid post.$outline-color;
```

File: packages/components/src/components/post-rating/post-rating.scss [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  z-index: 1;
+}
+
+.star-container {
+  position: relative;
+  display: block;
+  width: post.$lead-font-size;
+  height: post.$lead-font-size;
+  padding: 0 post.$size-micro 0 post.$size-micro;
+  cursor: pointer;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.scss:34]: You could use the approch to declair css variables here for fill and border, use them in `.fill` and `.border` classes and then only switch the css variables, when a modifier class (`.active-star`, `.hovered-star`, etc.) is appended to the `.star-container` element.

Also nest `.fill` and `.border` rules inside of the `.star-container` to make clear, they need to be a child of it.

```scss
.star-container {
  ...
  --post-rating-star-fill-color: #{post.$white};
  --post-rating-star-border-color: #{post.$black};
  
  .fill {
    position: absolute;
    color: var(--post-rating-star-fill-color);
  }
  
  .border {
    position: absolute;
    ...
    color: var(--post-rating-star-border-color);
  }
  
  &.active-star {
    --post-rating-star-fill-color: #{post.$yellow};
    --post-rating-star-border-color: #{post.$black};
  }
  
  ...
}
```

File: packages/components/src/components/post-rating/post-rating.scss [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  outline: post.$input-focus-outline-thickness solid post.$input-focus-color;
+  position: relative;
+  z-index: 1;
+}
+
+.star-container {
+  position: relative;
+  display: block;
+  width: post.$lead-font-size;
+  height: post.$lead-font-size;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.scss:32]: Why not using the post.$size-{x} variables? Or just add `1em`, as this is the current outputted value.

File: packages/components/src/components/post-rating/post-rating.scss [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+.rating:focus-visible {
+  outline-offset: post.$input-focus-outline-thickness;
+  outline: post.$input-focus-outline-thickness solid post.$input-focus-color;
+  position: relative;
+  z-index: 1;
+}
+
+.star-container {
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.scss:28]:
```suggestion
.star {
```

Keep it simple ;)

File: packages/components/src/components/post-rating/post-rating.scss [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  position: absolute;
+  color: post.$white;
+}
+
+.border {
+  width: post.$lead-font-size + 0.125em;
+  height: post.$lead-font-size + 0.125em;
+  position: absolute;
+  top: -0.0625rem;
+  left: post.$size-micro - 0.0625rem;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.scss:96]: Because you ussed `em` for `width` and `height`, you should also use `em` here, instead of `rem`, to make sure, it will always offset dependant on the current `font-size`.

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+  /**
+   * The index of the currently hovered star
+   */
+  @State() hoveredIndex: number;
+
+  /**
+   * The number of stars in the rating
+   */
+  @Prop() readonly max?: number = 5;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.tsx:24]: Misleading prop name. As you mention in the JSDocs comment, this is the number (or amount) of renderd stars, not the max count of'em. You could call it `stars`, `amount`, `count`, etc.
And please define some min (3) and max (10) values for it.
https://storybook.js.org/docs/api/arg-types#controlmax

Also move this prop up and implement it as the first prop, because it is the most common prop to use on this web-component.

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        break;
+      default:
+        return;
+    }
+  }
+
+  private hasChanged = false;
+
+  private handleBlur() {
+    if (this.hasChanged === true) {
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.tsx:93]: What if it is disabled or readonly? Is it working as expected?

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+      this.ratingChange.emit(this.currentRating);
+      this.hasChanged = false;
+    }
+  }
+
+  private handleHover(index: number, e: MouseEvent) {
+    if (!this.isInteractive()) return;
+    if (e.type === 'mouseenter') {
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.tsx:101]: Are you sure, you don't need the `mouseover` instead of the `mouseenter` event here?

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  private isInteractive(): boolean {
+    return !this.readonly && !this.disabled;
+  }
+
+  private update(value: number): void {
+    if (!this.isInteractive()) return;
+    if (value > this.max || value < 0) return;
+    this.currentRating = this.currentRating !== value ? value : 0;
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.tsx:115]: Would be nice to add a comment here, to explain, we are resetting the currentRating, if a users clicks the same star again.
I'm also not sure if this is a good and wanted behavior?
Let's discuss this with the team.

File: packages/components/src/index.html [hunk 13]
```diff
@@ -7,7 +7,6 @@
       content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
     />
     <title>Stencil Component Starter</title>
-
```
Comment (oliverschuerch) [inline packages/components/src/index.html:10]: please reset this unnecessary change

File: packages/documentation/src/stories/components/rating/post-rating.docs.mdx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+### Readonly
+
+Is the readonly property set, is the component not interactive
+but the current rating is still visible as in the default state.
+
+<Canvas of={RatingStories.Readonly} />
+<div className="hide-col-default">
+  <Controls of={RatingStories.Readonly} />
+</div>
```
Comment (oliverschuerch) [inline packages/documentation/src/stories/components/rating/post-rating.docs.mdx:44]: This is already explained and handled in the default story. If at all, I would add examples, which show how to event listener to the given events. Otherwise, I would remove this duplicated examples.

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+   * If readonly is true, the component only displays a rating and is not interactive.
+   */
+  @Prop() readonly readonly?: boolean = false;
+
+  // prettier-ignore
+  /**
+   * Event emitted when the rating gets commited
+   */
+  @Event({
+    eventName: 'ratingChange',
```
Comment (oliverschuerch) [inline packages/components/src/components/post-rating/post-rating.tsx:41]:
```suggestion
    eventName: 'change',
```

Yes it may show a warning in the terminal, but still works and is much closer to the standard.
Check if everything works as expeted, after changing this!

File: packages/documentation/src/stories/components/rating/post-rating.stories.ts [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+      name: 'disabled',
+      control: { type: 'boolean' },
+      description: 'Boolean for the disabled state of the component',
+      defaultValue: false,
+    },
+    readonly: {
+      name: 'readonly',
+      control: { type: 'boolean' },
+      description:
+        'If readonly is true, the component only displays a rating and is not interactive.',
```
Comment (oliverschuerch) [inline packages/documentation/src/stories/components/rating/post-rating.stories.ts:37]:
```suggestion
        'If readonly is `true`, the component only displays a rating and is not interactive.',
```

File: packages/documentation/src/stories/components/rating/post-rating.stories.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+export const CurrentRating: Story = {
+  parameters: {
+    controls: {
+      include: ['currentRating'],
+    },
+  },
+  args: {
+    currentRating: 3,
+  },
+};
```
Comment (oliverschuerch) [inline packages/documentation/src/stories/components/rating/post-rating.stories.ts:88]: Do you really use this story somewhere?

**Review by oliverschuerch — CHANGES_REQUESTED** [review]

File: packages/components/src/components/post-rating/post-rating.scss (hunk 6 above)
Comment (davidritter-dotcom) [reply packages/components/src/components/post-rating/post-rating.scss:32]: I could also just use 1em, the goal is that the stars have the same size as the text.

File: packages/components/src/components/post-rating/post-rating.tsx (hunk 10 above)
Comment (davidritter-dotcom) [reply packages/components/src/components/post-rating/post-rating.tsx:93]: I think so, because `hasChanged` can only get `true` if `isInteractive` is true.

File: packages/components/package.json [hunk 18]
```diff
@@ -28,7 +28,7 @@
     "start": "stencil build --watch --docs-readme",
     "build": "stencil build --docs-readme",
     "clean": "rimraf www dist loader",
-    "test": "npm run unit",
+    "test": "pnpm run unit",
```
[the remark this answers is not in the document]
Comment (davidritter-dotcom) [reply packages/components/package.json:31]: This does not necessarily need to be a part of this PR, I just encountered an error with the pnpm bootstrap, and it could be fixed with using pnpm instead of npm.

File: packages/components/cypress/e2e/rating.cy.ts (hunk 3 above)
Comment (davidritter-dotcom) [reply packages/components/cypress/e2e/rating.cy.ts:34]: I already tried the mouseover event, and it also doesn't work.

File: packages/components/src/components/post-rating/post-rating.tsx (hunk 11 above)
Comment (davidritter-dotcom) [reply packages/components/src/components/post-rating/post-rating.tsx:101]: I tried the mouseover event, but it kind of messes with my logic, maybe because I have two stars on top of each other.

File: packages/components/src/components/post-rating/post-rating.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliverschuerch) [reply packages/components/src/components/post-rating/post-rating.tsx]: We could do this with another PR.

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+})
+export class PostRating {
+  private hasChanged = false;
+
+  @Element() host: HTMLPostRatingElement;
+
+  @State() hoveredIndex: number;
+
+  /**
+   * Defined the invisible label on the component.
```
Comment (alizedebray) [inline packages/components/src/components/post-rating/post-rating.tsx:17]:
```suggestion
   * Defines an hidden label for the component.
```

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  @Prop() readonly label: string = 'Rating';
+
+  /**
+   * Defined the amount of stars rendered in the component.
```
Comment (alizedebray) [inline packages/components/src/components/post-rating/post-rating.tsx:22]:
```suggestion
   * Defines the total amount of stars rendered in the component.
```

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  @Prop() readonly stars: number = 5;
+
+  /**
+   * Defines the visualized overall rating, the component should show.
```
Comment (alizedebray) [inline packages/components/src/components/post-rating/post-rating.tsx:27]:
```suggestion
   * Defines the rating that the component should show.
```

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+  /**
+   * Defines if the component is readonly or not.
+   * This usually should be used together with the `currentRating` property.
+   */
+  @Prop() readonly readonly: boolean = false;
+
+  // prettier-ignore
+  /**
+   * An event emitted whenever the components value has changed (on input).
```
Comment (alizedebray) [inline packages/components/src/components/post-rating/post-rating.tsx:39]:
```suggestion
   * An event emitted whenever the component's value has changed (on input).
```

File: packages/components/src/components/post-rating/post-rating.tsx [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+   * The event payload can be used like so: `event.detail.value`.
+   */
+  @Event() input: EventEmitter<{value: number}>;
+
+  // prettier-ignore
+  /**
+   * An event emitted whenever the components value has changed (on blur).
```
Comment (alizedebray) [inline packages/components/src/components/post-rating/post-rating.tsx:46]:
```suggestion
   * An event emitted whenever the component's value has changed (on blur).
```

File: packages/documentation/src/stories/components/rating/post-rating.docs.mdx [hunk 24]
```diff
@@ -0,0 +1,31 @@
+import { Canvas, Controls, Meta, Source } from '@storybook/blocks';
+import LinkTo from '@storybook/addon-links/react';
+import * as RatingStories from './post-rating.stories';
+import SampleEvents from './post-rating.events.sample?raw';
+
+<Meta of={RatingStories} />
+
+# Rating
+
+Our Rating Web Component allows you to easily integrate
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.docs.mdx:10]:
```suggestion
<p class="lead">Our Rating Web Component allows you to easily integrate 
```

File: packages/documentation/src/stories/components/rating/post-rating.docs.mdx [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+import * as RatingStories from './post-rating.stories';
+import SampleEvents from './post-rating.events.sample?raw';
+
+<Meta of={RatingStories} />
+
+# Rating
+
+Our Rating Web Component allows you to easily integrate 
+a customizable star rating system, providing an 
+interactive and visually appealing way to gather and display user feedback.
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.docs.mdx:12]:
```suggestion
interactive and visually appealing way to gather and display user feedback.</p>
```

File: packages/documentation/src/stories/components/rating/post-rating.docs.mdx [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+
+<Canvas of={RatingStories.Default} sourceState='shown'/>
+<div className="hide-col-default">
+  <Controls of={RatingStories.Default} />
+</div>
+
+## Installation
+
+The `<post-rating/>` element is part of the `@swisspost/design-system-components` package.
+For more information, read the <LinkTo kind="getting-started-components" story="docs">getting started with components guide</LinkTo>.
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.docs.mdx:22]:
```suggestion
For more information, read the <LinkTo kind="edfb619b-fda1-4570-bf25-20830303d483" story="docs">getting started with components guide</LinkTo>.
```

File: packages/documentation/src/stories/components/rating/post-rating.events.sample.ts [hunk 27]
```diff
@@ -0,0 +1,9 @@
+const rating = document.querySelector('post-rating') as HTMLPostRatingElement;
+
+rating.addEventListener('input', event => {
+  // use event.detail.value for what ever you like
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.events.sample.ts:4]:
```suggestion
  // use event.detail.value for whatever you like
```

File: packages/documentation/src/stories/components/rating/post-rating.events.sample.ts [hunk 28]
```diff
@@ -0,0 +1,9 @@
...
+});
+
+rating.addEventListener('change', event => {
+  // use event.detail.value for what ever you like
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.events.sample.ts:8]:
```suggestion
  // use event.detail.value for whatever you like
```

File: packages/documentation/src/stories/components/rating/post-rating.stories.ts [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+import { html, nothing } from 'lit';
+import { BADGE } from '../../../../.storybook/constants';
+
+const meta: Meta = {
+  id: '956e063b-b40c-4fe4-bc27-53b8c4ab1e81',
+  title: 'Components/Rating',
+  component: 'post-rating',
+  render: render,
+  parameters: {
+    badges: [BADGE.NEEDS_REVISION, BADGE.SINCE_V1],
```
Comment (alizedebray) [inline packages/documentation/src/stories/components/rating/post-rating.stories.ts:11]:
```suggestion
    badges: [],
```

**Review by alizedebray — APPROVED** [review]

**Review by oliverschuerch — APPROVED** [review]
